### DISCUSSION TITLE
Allow mutating underlying session from `AuthSession`

### DISCUSSION THREAD

**Discussion by johnbchron** [maxcountryman/axum-login#193] [General]
Hey there! Firstly, awesome crate! I love the decoupling and system boundaries that you've established here.

My use case is that I want to do a simple "Remember me" checkbox. As far as I can tell, that would require mutating the `tower_sessions::Session` directly, which I can't do with `AuthSession`. I can see that `AuthSession` tracks and mutates the `tower_sessions::Session`, but doesn't allow the user to do so.

I think it could be useful to mutate the session for changing session expiry, but also for any other session-based values that `tower_sessions` allows. I assume that it would be possible to fetch the session directly after I log in a user and then mutate it, but I'm not super confident in my understanding of how `tower_sessions` handles mutations.

Anyways, mutating the session from `AuthSession` would be awesome, and in the meantime please advise how to change the session expiry just after logging in a user.

Here's my current (simplified) login function:
```rust
let mut auth_session = use_context::<auth::AuthSession>()
  .ok_or_else(|| ServerFnError::new("Failed to get auth session"))?;

let user = match auth_session.authenticate(creds.clone()).await {
  Ok(Some(user)) => user,
  Ok(None) => return Ok(false),
  Err(e) => {
    return Err(ServerFnError::new(format!("Failed to authenticate: {e}")))
  }
};

auth_session
  .login(&user)
  .await
  .map_err(|e| ServerFnError::new(format!("Failed to log in: {e}")))?;

tracing::info!("logged in user: {} ({})", user.name, user.id.0);
```

It would be awesome if I could do something like
```rust
let inner_session = auth_session.inner_session_mut();
inner_session.set_expiry(Some(...));
```

**Comment by maxcountryman** [comment]
If you're using `axum`, then you can also include `Session` as an extractor in your handler. For example:

```rust
async fn my_handler(auth_session: AuthSession, session: Session) -> impl IntoResponse {
    // Now we can use both the auth session and backing session together in our handler here.
}
```

**Reply by johnbchron** [reply]
And I don't have to worry about any overlapping writes if I login a user using the `AuthSession` in the same function context?

**Reply by maxcountryman** [reply]
It's pointing to the same memory and `axum-login` is managing its own namespace in the session, so you won't overwrite that unless you deliberately overwrite the key `axum-login` manages (note that you can configure this key to be whatever you like if that's a concern for you).

**Reply by johnbchron** [reply]
Awesome, thanks!

**Reply by nmabhinandan** [reply]
The data struct which is being inserted into session (see below) is not public so will we have to serialise the session data to json on every request using `session.get_value` and insert it again or is there a simpler way to do this? 

```rust
// from session.rs L#41

#[derive(Debug, Clone, Deserialize, Serialize)]
struct Data<UserId> {
    user_id: Option<UserId>,
    auth_hash: Option<Vec<u8>>,
}
```

**Reply by maxcountryman** [reply]
You should not be manually altering session data managed by axum-login; only axum-login should ever mutate data stored under its keys and users should only interact with these stored values via the public API axum-login provides.

However, you can and should insert your own data into the session under separate keys.

**Reply by nmabhinandan** [reply]
That's what I thought. But I don't see any tower-sessions api to update expiry of a session by key. I just want to extend the expiry of auth session on every request.

**Reply by maxcountryman** [reply]
I'm not following what you mean "expiry by session key" and it's pretty clearly off topic for this thread and perhaps axum-login itself.
